pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1105 @@
+				// Requeue the failed node for update again.
+				r.nodeCIDRUpdateChannel <- workItem
+			}
+		case <-stopChan:
+			return
+		}
+	}
+}
+
+func (r *multiCIDRRangeAllocator) cccWorker() {
+	for r.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (r *multiCIDRRangeAllocator) processNextWorkItem() bool {
+	cKey, quit := r.queue.Get()
+	if quit {
+		return false
+	}
+	defer r.queue.Done(cKey)
+
+	err := r.syncClusterCIDRConfig(cKey.(string))
+	r.handleErr(err, cKey)
+
+	return true
+}
+
+func (r *multiCIDRRangeAllocator) handleErr(err error, key interface{}) {
+	if err == nil {
+		r.queue.Forget(key)
+		return
+	}
+
+	if r.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error syncing clusterCIDRConfig %q, retrying. Error: %v", key, err)
+		r.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Retry budget exceeded, dropping clusterCIDRConfig %q out of the queue: %v", key, err)
+	r.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (r *multiCIDRRangeAllocator) syncClusterCIDRConfig(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing clusterCIDRConfig request %q (%v)", key, time.Since(startTime))
+	}()
+
+	clusterCIDRConfig, err := r.clusterCIDRConfigLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		klog.V(3).Infof("clusterCIDRConfig has been deleted: %v", key)
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// Check the DeletionTimestamp to determine if object is under deletion.
+	if !clusterCIDRConfig.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(clusterCIDRConfig)
+	}
+	return r.reconcileCreate(clusterCIDRConfig)
+}
+
+func (r *multiCIDRRangeAllocator) insertNodeToProcessing(nodeName string) bool {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	if r.nodesInProcessing.Has(nodeName) {
+		return false
+	}
+	r.nodesInProcessing.Insert(nodeName)
+	return true
+}
+
+func (r *multiCIDRRangeAllocator) removeNodeFromProcessing(nodeName string) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	klog.Infof("Removing Node: %s from processing", nodeName)
+	r.nodesInProcessing.Delete(nodeName)
+}
+
+// marks node.PodCIDRs[...] as used in allocator's tracked cidrSet.
+func (r *multiCIDRRangeAllocator) occupyCIDRs(node *v1.Node) error {
+	defer r.removeNodeFromProcessing(node.Name)
+	if len(node.Spec.PodCIDRs) == 0 {
+		return nil
+	}
+
+	clusterCIDRList, err := r.orderedMatchingClusterCIDRs(node)
+	if err != nil {
+		return err
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	for _, clusterCIDR := range clusterCIDRList {
+		occupiedCnt := 0
+
+		for _, cidr := range node.Spec.PodCIDRs {
+			_, podCIDR, err := netutil.ParseCIDRSloppy(cidr)
+			if err != nil {
+				return fmt.Errorf("failed to parse CIDR %s on Node %v: %w", cidr, node.Name, err)
+			}
+
+			klog.Infof("occupy CIDR %s for node: %s", cidr, node.Name)
+
+			if err := r.Occupy(clusterCIDR, podCIDR); err != nil {
```

Comment:
What happens if some of the CIDRs succeed in Occupy, but then it fails in the middle? Do we release the ones that have already been occupied? If a Node had a list of CIDRs where some succeeded but then the last one fails, do we end up eating up all of the CIDRs for the cluster via the retry?

Reply:
If the Node has some CIDRs already assigned, we should mark them occupied and those shouldn't be used for allocation any more, in the case where occupy succeeds for some, we are correctly eating up the CIDRs, however the issue would be for the failed CIDR where we might end up allocating the used CIDR to some other node

A similar issue exists in the existing RangeAllocator as well, MultiCIDRRangeAllocator implementation is based on the existing RangeAllocator implementation, with an additional support for Multiple ClusterCIDRs